File: kernel/coc/recall.py
"""`table.recall` — the transcript and history roads (contract §12.4). The memory road
lives in memory.py next to the store it reads.

Transcript: structured selectors only (turn range, role, an exact {turn, role} locator).
`verified` is the sha256 of the transcript row's text against the turn record's own
copy (`rendered_text` for the keeper, `player_text` for the player): a transcript row
that drifted from what the turn record says is returned anyway, marked false.

History: timeline and diff come from `turns/NNNN.json` alone, never from git objects."""

from __future__ import annotations

from typing import Any

from .errors import invalid_params, unsupported_value
from .events import EVENT_TYPES
from .facts import head_of
from .fileio import sha256_text
from .store import Campaign

TRANSCRIPT_DEFAULT_TURNS = 3
TRANSCRIPT_MAX_CARDS = 40
CARD_HEAD_CHARS = 80
HISTORY_DEFAULT_TURNS = 20
HISTORY_MAX_EVENTS = 200
RECEIPT_KINDS = ("roll", "move", "clue", "delta", "session", "time")
ROLES = ("player", "keeper")


def parse_span(value: Any, current: int, default_turns: int) -> list[int]:
    if value is None:
        return [max(0, current - default_turns + 1), current]
    if (not isinstance(value, list) or len(value) != 2
            or not all(isinstance(v, int) and not isinstance(v, bool) and v >= 0 for v in value) or value[0] > value[1]):
        raise invalid_params("turns must be [from, to] with 0 <= from <= to")
    return [int(value[0]), int(value[1])]


def parse_role(value: Any) -> str | None:
    if value is not None and value not in ROLES:
        raise invalid_params("role must be 'player' or 'keeper'")
    return value


# ---- transcript ---------------------------------------------------------------------------

def record_text(record: dict[str, Any] | None, role: str) -> str | None:
    if record is None:
        return None
    return record.get("rendered_text") if role == "keeper" else record.get("player_text")


def transcript(campaign: Campaign, current: int, params: dict[str, Any]) -> dict[str, Any]:
    read = params.get("read")
    if read is not None:
        return transcript_read(campaign, read)
    span = parse_span(params.get("turns"), current, TRANSCRIPT_DEFAULT_TURNS)
    role = parse_role(params.get("role"))
    entries = [e for e in campaign.read_transcript()
               if span[0] <= int(e["turn"]) <= span[1] and (role is None or e["role"] == role)]
    cards = [{"turn": int(e["turn"]), "role": e["role"], "chars": len(e.get("text") or ""),
              "head": " ".join(str(e.get("text") or "").split())[:CARD_HEAD_CHARS]}
             for e in entries[-TRANSCRIPT_MAX_CARDS:]]
    result: dict[str, Any] = {"what": "transcript", "turns": span, "cards": cards}
    if span[1] - span[0] + 1 <= TRANSCRIPT_DEFAULT_TURNS:
        # slice 0 behaviour kept for short windows: the rows themselves
        result["entries"] = entries
    return result


def transcript_read(campaign: Campaign, read: Any) -> dict[str, Any]:
    if (not isinstance(read, dict) or not isinstance(read.get("turn"), int) or isinstance(read.get("turn"), bool)
            or read.get("role") not in ROLES):
        raise invalid_params("read must be {turn: int, role: player|keeper}")
    turn, role = int(read["turn"]), str(read["role"])
    rows = [e for e in campaign.read_transcript() if int(e["turn"]) == turn and e["role"] == role]
    if not rows:
        available = sorted({(int(e["turn"]), e["role"]) for e in campaign.read_transcript()})
        raise invalid_params(f"no {role} row for turn {turn} in the transcript",
                             fix="pick a card from recall transcript first",
                             details={"turn": turn, "role": role,
                                      "available": [{"turn": t, "role": r} for t, r in available[-TRANSCRIPT_MAX_CARDS:]]})
    text = str(rows[-1].get("text") or "")
    canonical = record_text(campaign.read_turn_record(turn), role)
    verified = canonical is not None and sha256_text(text) == sha256_text(str(canonical))
    return {"what": "transcript", "turn": turn, "role": role, "text": text, "verified": verified}


# ---- history ------------------------------------------------------------------------------

def timeline_row(record: dict[str, Any]) -> dict[str, Any]:
    snapshot = record.get("world") if isinstance(record.get("world"), dict) else {}
    scene = snapshot.get("scene") or {}
    counts = {kind: 0 for kind in RECEIPT_KINDS}
    for receipt in record.get("receipts") or []:
        kind = receipt.get("kind")
        if kind in counts:
            counts[kind] += 1
    return {
        "turn": int(record["turn"]),
        "commit": record.get("commit"),
        "scene": scene.get("name"),
        "clock": (snapshot.get("clock") or {}).get("minutes"),
        "closed_by": record.get("closed_by"),
        "receipts": counts,
        "head": head_of(record.get("rendered_text")),
    }


def history(campaign: Campaign, current: int, params: dict[str, Any]) -> dict[str, Any]:
    span = parse_span(params.get("turns"), current, HISTORY_DEFAULT_TURNS)
    types = params.get("types")
    if types is not None:
        if not isinstance(types, list) or not all(isinstance(t, str) for t in types):
            raise invalid_params("types must be a list of event types")
        unknown = sorted(set(types) - EVENT_TYPES)
        if unknown:
            raise unsupported_value("types", sorted(unknown), sorted(EVENT_TYPES),
                                    message=f"unknown event types {unknown}")
    records = campaign.turn_records_by_number()
    timeline = [timeline_row(records[n]) for n in sorted(records) if span[0] <= n <= span[1]]
    events = [e for e in campaign.read_events()
              if span[0] <= int(e.get("turn", -1)) <= span[1] and (types is None or e.get("type") in types)]
    result: dict[str, Any] = {"what": "history", "turns": span, "timeline": timeline,
                              "events": events[-HISTORY_MAX_EVENTS:]}
    diff = params.get("diff")
    if diff is not None:
        result["diff"] = history_diff(records, diff)
    return result


def worldline_tree(meta: dict[str, Any]) -> dict[str, Any]:
    """§15.6: `recall history {lines: true}`. Every line the campaign has, in fork order:
    where it branched, which circuit of the loop it is, how far it played and whether it
    is the one at the table. Read from the registry alone; no git object is opened."""
    from .worldline import active_name, registry  # local: worldline reads history, which reads nothing here
    lines = registry(meta)
    active = active_name(meta)
    rows = []
    for name in sorted(lines):
        row = lines[name] if isinstance(lines[name], dict) else {}
        forked = row.get("forked_from") if isinstance(row.get("forked_from"), dict) else None
        rows.append({"name": name, "kind": row.get("kind"), "loop": int(row.get("loop") or 0),
                     "status": row.get("status"), "last_turn": row.get("last_turn"),
                     "last_commit": row.get("last_commit"),
                     "forked_from": ({"line": forked.get("line"), "turn": forked.get("turn"),
                                      "commit": forked.get("commit")} if forked else None),
                     "parents": row.get("parents") or [],
                     "active": name == active})
    return {"active": active, "lines": rows}


def history_diff(records: dict[int, dict[str, Any]], diff: Any) -> dict[str, Any]:
    """State at the end of turn a against the end of turn b, accumulated from the receipts
    of turns a+1..b. Only turn records are read."""
    if (not isinstance(diff, list) or len(diff) != 2
            or not all(isinstance(v, int) and not isinstance(v, bool) and v >= 0 for v in diff) or diff[0] > diff[1]):
        raise invalid_params("diff must be [turn_a, turn_b] with 0 <= turn_a <= turn_b")
    a, b = int(diff[0]), int(diff[1])
    missing = [n for n in (a, b) if n not in records]
    if missing:
        raise invalid_params(f"no turn record for {missing}", details={"closed_turns": sorted(records)})

    def snapshot(n: int) -> dict[str, Any]:
        world = records[n].get("world")
        return world if isinstance(world, dict) else {}

    clues: list[str] = []
    resources: dict[tuple[str, str], dict[str, Any]] = {}
    sessions: list[dict[str, Any]] = []
    moves: list[dict[str, Any]] = []
    for n in sorted(records):
        if not a < n <= b:
            continue
        for receipt in records[n].get("receipts") or []:
            kind = receipt.get("kind")
            if kind == "clue":
                name = receipt.get("clue")
                if name not in clues:
                    clues.append(str(name))
            elif kind in ("delta", "cash"):
                # #19: a cash receipt carries resource/subject/before/after like a delta
                key = (str(receipt.get("subject")), str(receipt.get("resource")))
                row = resources.setdefault(key, {"subject": key[0], "resource": key[1], "from": receipt.get("before"),
                                                 "to": receipt.get("after")})
                row["to"] = receipt.get("after")
            elif kind == "session":
                row = {"turn": n, "family": receipt.get("family"), "transition": receipt.get("transition")}
                if receipt.get("outcome") is not None:
                    row["outcome"] = receipt.get("outcome")
                sessions.append(row)
            elif kind == "move":
                moves.append({"turn": n, "from": receipt.get("from"), "to": receipt.get("to")})
    return {
        "from": a, "to": b,
        "scene": [(snapshot(a).get("scene") or {}).get("name"), (snapshot(b).get("scene") or {}).get("name")],
        "clock": [(snapshot(a).get("clock") or {}).get("minutes"), (snapshot(b).get("clock") or {}).get("minutes")],
        "clues_added": clues,
        "resources": list(resources.values()),
        "sessions": sessions,
        "moves": moves,
    }

File: kernel-ts/memory/recall.ts
/** Structured recall over retained records, transcripts and candidate sources. */
import { RpcError } from '../errors.js';
import { isJsonObject, parsePythonJson, sha256Text } from '../json.js';
import { CampaignWriter, EVENT_TYPES } from '../write/store.js';
import { ModuleGraph } from '../read/module-graph.js';
import { EntityIndex, queryCandidates, withPromiseFulfillment, canonicalMemoryReceipts, memoryOccurrenceKey } from '../read/memory.js';
import {lineFulfillmentEvidence} from '../read/worldline.js';
import { npcsPresent } from '../read/capsule.js';
import { array, row, number, integer, string, truth, sorted, chars, length, words, repr, type Row } from '../read/values.js';
import { CANDIDATE_KINDS, logs, records, proseOf } from './jobs.js';
import {RECALL_CHARS, pageOptions, detailOptions, position, publicRecall, recallBytes, recallSnapshot, textPage, rowsPage, rowDetail} from './pages.js';
const ROLES = ['player', 'keeper'];
const RECEIPTS = ['roll', 'move', 'clue', 'delta', 'session', 'time', 'clock'];
export function parseSpan(value: any, current: number, defaults: number): number[] {
    if (value == null)
        return [Math.max(0, current - defaults + 1), current];
    if (!Array.isArray(value) || value.length !== 2 || !value.every(n => integer(n) && number(n) >= 0) || number(value[0]) > number(value[1]))
        throw new RpcError('invalid_params', 'turns must be [from, to] with 0 <= from <= to');
    return value.map(number);
}
export async function transcript(campaign: CampaignWriter, current: number, params: Row): Promise<Row> {
    const all = await logs(campaign, 'transcript.jsonl'), read = params.read;
    const snapshot = await recallSnapshot(campaign, params, all);
    if (read != null) {
        if (!isJsonObject(read) || Object.keys(read).some(key => !['turn', 'role', 'offset', 'limit'].includes(key))
            || !integer(read.turn) || number(read.turn) < 0 || !ROLES.includes(read.role as string))
            throw new RpcError('invalid_params', 'read must name a turn, role and optional offset/limit');
        const turn = number(read.turn), role = string(read.role), found = all.filter(value => number(value.turn) === turn && value.role === role);
        if (!found.length)
            throw new RpcError('invalid_params', `no ${role} row for turn ${turn} in the transcript`, {
                fix: 'Pick a card from recall transcript first.', details: {turn, role},
            });
        const text = string(found.at(-1)!.text || ''), record = await campaign.readTurnRecord(turn);
        const canonical = record?.[role === 'keeper' ? 'rendered_text' : 'player_text'];
        const offset = position(read.offset, 0, 'read.offset'), limit = position(read.limit, RECALL_CHARS, 'read.limit', true);
        return textPage(text, offset, limit, {what: 'transcript', turn, role, _snapshot: snapshot,
            verified: canonical != null && sha256Text(text) === sha256Text(string(canonical)), verification_scope: 'record_integrity_only'},
            offset => ({what: 'transcript', read: {turn, role, offset, limit}}));
    }
    const span = parseSpan(params.turns, current, 3), role = params.role ?? null;
    if (role !== null && !ROLES.includes(role))
        throw new RpcError('invalid_params', "role must be 'player' or 'keeper'");
    const latest = new Map<string, Row>();
    for (const value of all) if (span[0] <= number(value.turn) && number(value.turn) <= span[1] && (role === null || value.role === role))
        latest.set(`${value.turn}:${value.role}`, value);
    const cards = [...latest.values()].sort((a, b) => number(a.turn) - number(b.turn) || ROLES.indexOf(a.role) - ROLES.indexOf(b.role))
        .map(value => ({turn: number(value.turn), role: value.role, chars: length(value.text || ''), head: chars(string(value.text || ''), 80),
            read: {what: 'transcript', read: {turn: number(value.turn), role: value.role, offset: 0, limit: RECALL_CHARS}}}));
    const query = {...params, turns: span}, base = {what: 'transcript', turns: span, _snapshot: snapshot};
    return rowDetail(cards, query, 'cards', base) ?? rowsPage(cards, query, 'cards', base);
}
function timelineRow(record: Row): Row {
    const world = row(record.world), counts = Object.fromEntries(RECEIPTS.map(kind => [kind, 0]));
    for (const receipt of array(record.receipts))
        if (RECEIPTS.includes(receipt.kind))
            counts[receipt.kind]++;
    return { turn: number(record.turn), commit: record.commit ?? null, scene: row(world.scene).name ?? null, clock: row(world.clock).minutes ?? null,
        closed_by: record.closed_by ?? null, receipts: counts, head: chars(words(proseOf(record.rendered_text)), 60) };
}
function historyDiff(all: Map<number, Row>, diff: any): Row {
    if (!Array.isArray(diff) || diff.length !== 2 || !diff.every(value => integer(value) && number(value) >= 0) || number(diff[0]) > number(diff[1]))
        throw new RpcError('invalid_params', 'diff must be [turn_a, turn_b] with 0 <= turn_a <= turn_b');
    const [a, b] = diff.map(number), missing = [a, b].filter(turn => !all.has(turn));
    if (missing.length)
        throw new RpcError('invalid_params', `no turn record for ${repr(missing)}`, { details: { closed_turns: [...all.keys()].sort((a, b) => a - b) } });
    const clues: string[] = [], resources = new Map<string, Row>(), sessions: Row[] = [], moves: Row[] = [];
    for (const [turn, value] of [...all].sort(([a], [b]) => a - b))
        if (a < turn && turn <= b)
            for (const receipt of array(value.receipts)) {
                if (receipt.kind === 'clue') {
                    if (!clues.includes(receipt.clue))
                        clues.push(string(receipt.clue));
                }
                else if (['delta', 'cash'].includes(receipt.kind)) {
                    const subject = string(receipt.subject), resource = string(receipt.resource), key = JSON.stringify([subject, resource]);
                    if (!resources.has(key))
                        resources.set(key, { subject, resource, from: receipt.before ?? null, to: receipt.after ?? null });
                    resources.get(key)!.to = receipt.after ?? null;
                }
                else if (receipt.kind === 'session')
                    sessions.push({ turn, family: receipt.family ?? null, transition: receipt.transition ?? null, ...(receipt.outcome != null ? { outcome: receipt.outcome } : {}) });
                else if (receipt.kind === 'move')
                    moves.push({ turn, from: receipt.from ?? null, to: receipt.to ?? null });
            }
    const left = row(all.get(a)!.world), right = row(all.get(b)!.world);
    return { from: a, to: b, scene: [row(left.scene).name ?? null, row(right.scene).name ?? null], clock: [row(left.clock).minutes ?? null, row(right.clock).minutes ?? null],
        clues_added: clues, resources: [...resources.values()], sessions, moves };
}
function worldlineTree(meta: Row): Row {
    const lines = row(meta.worldlines), active = typeof meta.active_worldline === 'string' && meta.active_worldline ? meta.active_worldline : 'main';
    return { active, lines: sorted(Object.keys(lines)).map(name => {
            const value = row(lines[name]), forked = row(value.forked_from);
            return { name, kind: value.kind ?? null, loop: number(value.loop), status: value.status ?? null, last_turn: value.last_turn ?? null, last_commit: value.last_commit ?? null,
                forked_from: truth(forked) ? { line: forked.line ?? null, turn: forked.turn ?? null, commit: forked.commit ?? null } : null, parents: truth(value.parents) ? value.parents : [], active: name === active };
        }) };
}
export async function history(campaign: CampaignWriter, current: number, params: Row): Promise<Row> {
    const span = parseSpan(params.turns, current, 20), types = params.types;
    if (types != null) {
        if (!Array.isArray(types) || !types.every(value => typeof value === 'string'))
            throw new RpcError('invalid_params', 'types must be a list of event types');
        const unknown = sorted(new Set(types.filter(value => !EVENT_TYPES.has(value))));
        if (unknown.length) throw new RpcError('invalid_params', 'History contains an unsupported event type', {
            fix: 'Use an event type supplied by the tool schema.', details: {field: 'types'},
        });
    }
    const all = await records(campaign), events = await logs(campaign, 'events.jsonl');
    const lines = truth(params.lines) ? worldlineTree(await campaign.readCampaign()) : null;
    const snapshot = await recallSnapshot(campaign, params, [[...all], events, lines]);
    const requested = detailOptions(params)?.section ?? pageOptions(params, params.diff != null ? 'diff' : types != null ? 'events' : 'timeline').section;
    const query = {...params, turns: span}, base: Row = {what: 'history', turns: span, _snapshot: snapshot};
    const timeline = [...all].sort(([a], [b]) => a - b).filter(([turn]) => span[0] <= turn && turn <= span[1]).map(([, value]) => timelineRow(value));
    const largeLines = lines != null && recallBytes(lines) > 2048;
    if (lines) base.lines = largeLines ? {active: chars(string(lines.active), 80), truncated: true,
        read: {...publicRecall(query), page: {section: 'timeline', offset: timeline.length, limit: 1}}} : lines;
    let selected: Row[];
    if (requested === 'events') selected = events.filter(value => span[0] <= number(value.turn, -1) && number(value.turn, -1) <= span[1] && (types == null || types.includes(value.type)));
    else if (requested === 'timeline') {
        selected = timeline;
        if (largeLines) selected.push({kind: 'worldlines', ...lines});
    } else if (requested === 'diff') {
        const diff = historyDiff(all, params.diff);
        base.from = diff.from; base.to = diff.to;
        selected = [{kind: 'scene', from: diff.scene[0], to: diff.scene[1]}, {kind: 'clock', from: diff.clock[0], to: diff.clock[1]},
            ...diff.clues_added.map((clue: string) => ({kind: 'clue', clue})),
            ...diff.resources.map((value: Row) => ({kind: 'resource', ...value})),
            ...diff.sessions.map((value: Row) => ({kind: 'session', ...value})),
            ...diff.moves.map((value: Row) => ({kind: 'move', ...value}))];
    } else throw new RpcError('invalid_params', 'History sections are timeline, events and diff');
    return rowDetail(selected, query, requested, base) ?? rowsPage(selected, query, requested, base);
}
async function lineCandidates(campaign: CampaignWriter, line: string): Promise<Row[]> {
    const raw = await campaign.context.git.lineBlob(campaign.id, line, 'memory/candidates.jsonl') || '', result: Row[] = [];
    for (const text of raw.split(/\r\n|[\n\r\v\f\x1c-\x1e\x85\u2028\u2029]/)) {
        if (!text.trim())
            continue;
        try {
            const value = parsePythonJson(text);
            if (isJsonObject(value))
                result.push(value);
        }
        catch { /* Unparseable rows in other retained lines remain absent from the view. */ }
    }
    const evidence=await lineFulfillmentEvidence(campaign,line);
    return withPromiseFulfillment(evidence.memory??result,{campaign:campaign.id,...evidence});
}
export async function candidatesFor(campaign: CampaignWriter, meta: Row, line: any): Promise<Row[]> {
    const current=async()=>{const [memory,world,turn,history]=await Promise.all([logs(campaign,'memory/candidates.jsonl'),campaign.readWorld(),campaign.readTurn(),records(campaign)]);
        return withPromiseFulfillment(memory,{campaign:campaign.id,receipts:canonicalMemoryReceipts([...history.values()],array(turn.receipts)),world});};
    if (line == null || line === 'current'||line===(meta.active_worldline||'main')) return current();
    const lines = row(meta.worldlines);
    if (line !== 'any') {
        if (typeof line !== 'string' || !Object.hasOwn(lines, line))
            throw new RpcError('invalid_params', `no worldline ${repr(line)}`, { fix: `one of 'current', 'any', or ${repr(sorted(Object.keys(lines)))}`, details: { line, lines: sorted(Object.keys(lines)) } });
        return lineCandidates(campaign, line);
    }
    const active = typeof meta.active_worldline === 'string' && meta.active_worldline ? meta.active_worldline : 'main', rows = new Map<string, Row>();
    for (const name of sorted(Object.keys(lines)))
        if (name !== active)
            for (const value of await lineCandidates(campaign, name))
                if (!rows.has(memoryOccurrenceKey(value)))
                    rows.set(memoryOccurrenceKey(value), value);
    for (const value of await current())
        rows.set(memoryOccurrenceKey(value), value);
    return sorted(rows.keys()).map(key => rows.get(key)!);
}
export async function recallMemory(campaign: CampaignWriter, graph: ModuleGraph, world: Row, params: Row): Promise<Row> {
    const party = await campaign.party(), index = new EntityIndex(graph, party, row(world.scene_labels)), narrow = params.about != null;
    let about: string[];
    if (!narrow)
        about = [...npcsPresent(graph, world, graph.scene(world.active_scene)).map(node => graph.displayName(node)), ...party.map(sheet => string(sheet.name))];
    else {
        if (!Array.isArray(params.about) || !params.about.every((value: any) => typeof value === 'string' && value.trim()))
            throw new RpcError('invalid_params', 'about must be a list of names');
        about = params.about.map((name: string) => {
            const exact = index.matches(name), found = exact.length ? exact : index.looseMatches(name);
            if (found.length !== 1) {
                const names = found.length ? found.map(key => index.canonicalName(key)) : graph.candidates(name, ['npc', 'scene', 'clue']);
                throw new RpcError('unknown_entity', `${repr(name)} is ${found.length ? 'ambiguous' : 'not a known name'}`, {
                    fix: names.length ? `use one of ${repr(names)}` : 'use an investigator, NPC, scene or clue name, or world/party/keeper/player', details: { query: name, candidates: names }
                });
            }
            return index.canonicalName(found[0]);
        });
    }
    const turns = params.turns == null ? undefined : parseSpan(params.turns, 0, 0), kinds = params.kinds;
    if (kinds != null && (!Array.isArray(kinds) || !kinds.every((kind: any) => CANDIDATE_KINDS.includes(kind))))
        throw new RpcError('invalid_params', 'kinds must be a list of candidate kinds', { fix: `one of ${repr(CANDIDATE_KINDS)}` });
    const limit = Object.hasOwn(params, 'limit') ? params.limit : 12;
    position(limit, 12, 'limit', true);
    const line = Object.hasOwn(params, 'line') ? params.line : 'current', rows = await candidatesFor(campaign, await campaign.readCampaign(), line);
    const hits = queryCandidates(rows, index, about, {narrow, turns, kinds, includeSuperseded: truth(params.include_superseded), limit: rows.length});
    const snapshot = await recallSnapshot(campaign, params, [rows, hits, line]);
    const query = {...params, page: {limit: number(limit), ...row(params.page)}};
    const base = {what: 'memory', about, line, _snapshot: snapshot};
    return rowDetail(hits, query, 'hits', base) ?? rowsPage(hits, query, 'hits', base);
}
